### DISCUSSION TITLE
export headers are different than what import headers require

### DISCUSSION THREAD

**Discussion by fvandonk** [netbox-community/netbox#13140] [Help Wanted!] [answered]
I am exporting data from my 3.5.4 netbox and importing them into my 3.5.5 netbox.
When I import the CSV files they all fail, when I look at the headers I notice that all columns have the first character capitalized.
Also some of the field do not even have the correct header field name.

Is this something that as changed between the two versions?
It is not a huge deal to change the headers but it seems a bit sloppy is all.
So far this is happening on everything in Organization, Devices, IPAM - IP Addresses)

**Comment by candlerb** [comment, marked as answer]
> I am exporting data from my 3.5.4 netbox and importing them into my 3.5.5 netbox.

In general, it's impossible to use CSV export and import to transfer data between Netbox instances with any sort of fidelity. There are many things which can't be done this way, especially anything which involves foreign key relations where the target model doesn't have a unique name. Even transferring a device with its primary_ip can't be done directly: you'd have to import the devices, then import the interfaces, then import the IP addresses, then update the devices to set their primary_ip(4/6). Note also that the last step would also involve mapping between the original device ID and the new device ID, since the only way to update via CSV involves supplying the device ID, but the import process assigns new database IDs.

You will be much better off simply doing a postgres [dump and restore](https://docs.netbox.dev/en/stable/administration/replicating-netbox/), and then re-running the `./upgrade.sh` script to apply any database schema changes (although I don't think there were any between 3.5.4 and 3.5.5).  Of course, you'll lose any data you had manually entered into 3.5.5.

> Is this something that as changed between the two versions?

No, it's not a change: CSV export has _never_ been compatible with CSV import. It's mentioned in the [FAQ](https://github.com/netbox-community/netbox/wiki/Frequently-Asked-Questions#why-cant-i-pass-exported-data-directly-into-the-import-form).

If you want to export CSV in a format which is approximately compatible with import (such as using the correct column headers), then you have to write your own [export templates](https://docs.netbox.dev/en/stable/customization/export-templates/).

Here's an example for exporting cables in a structure for editing and reimporting - #11762, #12018. (Although that was before cables supported multiple terminations, and I'm not sure that you can import those via CSV).

**Reply by fvandonk** [reply]
Thanks for taking the time to reply and provide me the details required to move forward. 
Even though I understand what is said in the FAQ I do not agree with it as it sounds more like an easy way out fixing a broken function. But that is just my 2ct. 
In my case I had to go this route because I had a demo setup with the container based solution and needed to get the data out to put it in my regular installation (non container). Instead of spending hours trying to figure out how to get the data backed up and out of the container I decided to go the export/import route.
Now I am working on testing the backup and restore functionality as it is an important part of evaluating a product before putting it into production. This besides VM based snapshot backups.
Thanks again.

**Reply by candlerb** [reply]
> Instead of spending hours trying to figure out how to get the data backed up and out of the container I decided to go the export/import route

It should be easy enough: see #9681 for clues.

It could be as simple as:

```
docker exec netbox-postgres-1 /usr/local/bin/pg_dump netbox -U netbox
```
